plugins/ssr/ssripats.ml

```diff
@@ -48,39 +64,91 @@ type istate = {
   (* Delayed clear *)
   to_clear : Id.t list;
 
+  (* Temporary intros, to be generalized back *)
+  to_generalize : delayed_gen list;
+
+  (* The type of the inductive constructor corresponding to the current proof
+   * branch: name seeds are taken from that in an intro block *)
+  name_seed : EConstr.types option;
+
+}
+and delayed_gen = {
+  tmp_id : Id.t;    (* Temporary name *)
+  orig_name : Name.t   (* Old name *)
 }
 
 let empty_state = {
   to_clear = [];
+  to_generalize = [];
+  name_seed = None;
 }
 
 include Ssrcommon.MakeState(struct
   type state = istate
   let init = empty_state
 end)
 
+let ns_print env sigma = function
+  | None -> Pp.str "-"
+  | Some t -> Printer.pr_econstr_env env sigma t
+
+let dg_print { tmp_id; orig_name } =
+  Pp.(Id.print tmp_id ++ str"->" ++ Name.print orig_name)
+
 let isPRINT g =
+  let env, sigma = Goal.env g, Goal.sigma g in
   let state = get g in
   Pp.(str"{{ to_clear: " ++
         prlist_with_sep spc Id.print state.to_clear ++ spc () ++
-      str" }}")
+      str"to_generalize: " ++
+        prlist_with_sep spc dg_print state.to_generalize ++ spc () ++
+      str"name_seed: " ++ ns_print env sigma state.name_seed ++ str" }}")
 
 
 let isCLR_PUSH id =
-  tclGET (fun { to_clear = ids } ->
-  tclSET { to_clear = id :: ids })
+  tclGET (fun ({ to_clear = ids } as s) ->
+  tclSET { s with to_clear = id :: ids })
 
 let isCLR_PUSHL more_ids =
-  tclGET (fun { to_clear = ids } ->
-  tclSET { to_clear = more_ids @ ids })
+  tclGET (fun ({ to_clear = ids } as s) ->
+  tclSET { s with to_clear = more_ids @ ids })
 
 let isCLR_CONSUME =
-  tclGET (fun { to_clear = ids } ->
-  tclSET { to_clear = [] } <*>
+  tclGET (fun ({ to_clear = ids } as s) ->
+  tclSET { s with to_clear = [] } <*>
   Tactics.clear ids)
 
 
-let isTICK _ = tclUNIT ()
+let isGEN_PUSH dg =
+  tclGET (fun s ->
+  tclSET { s with to_generalize = dg :: s.to_generalize })
+
+let gen_astac id new_name =
+ let gen = ((None,Some(false,[])),Ssrmatching.cpattern_of_id id) in
+ V82.tactic (Ssrcommon.gentac gen)
+ <*> Ssrcommon.tclRENAME_HD_PROD new_name
+
+let isGEN_CONSUME =
+  tclGET (fun ({ to_generalize = dgs } as s) ->
+  tclSET { s with to_generalize = [] } <*>
+  Tacticals.New.tclTHENLIST
+    (List.map (fun { tmp_id; orig_name } ->
+       gen_astac tmp_id orig_name) dgs) <*>
+  Tactics.clear (List.map (fun gen -> gen.tmp_id) dgs))
+
+
+let isNSEED_SET ty =
+  tclGET (fun s ->
+  tclSET { s with name_seed = Some ty })
+
+let isNSEED_CONSUME k =
+  tclGET (fun ({ name_seed = x } as s) ->
+  tclSET { s with name_seed = None } <*>
+  k x)
+
+let isTICK = function
```

Comment:
I don't understand the name. Why tick?

Reply:
Tick is called between ipats and makes the name seed ephemeral: if seeding is followed by a block intro is fine, else we discard the seed.

Reply:
comment added